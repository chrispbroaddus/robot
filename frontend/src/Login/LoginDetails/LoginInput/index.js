import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

const LoginField = styled.input`
  width: 210px;
  height: 30px;
  padding: 5px 10px;
  margin: 0 auto;
  margin-bottom: 10px;
  border: 1px solid #cfd8dc;
  border-radius: 5px;
  color: #90a4ae;
  font-size: 0.8rem;
  display: block;

  &::placeholder {
    color: #90a4ae;
    font-family: Montserrat-Light;
    letter-spacing: 0.5px;
  }
`

const LoginInput = props => {
  return (
    <LoginField
      name={props.name}
      type={props.type}
      value={props.value}
      onChange={props.handleChange}
      placeholder={props.placeholder}
      innerRef={props.refFunc}
    />
  )
}

LoginInput.propTypes = {
  value: PropTypes.string,
  placeholder: PropTypes.string,
  handleChange: PropTypes.func.isRequired,
  name: PropTypes.string,
  type: PropTypes.string,
  // Using the property "refFunc" because React complains when we try to use ref
  refFunc: PropTypes.func
}

LoginInput.defaultProps = {
  type: 'text'
}

export default LoginInput
