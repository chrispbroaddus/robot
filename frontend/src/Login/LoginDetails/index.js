import React from 'react'
import PropTypes from 'prop-types'
import styled, { keyframes } from 'styled-components'
import LoginInput from './LoginInput'

const Container = styled.div`
  position: relative;
  margin: auto;
  margin-top: 250px;
  text-align: center;
  padding: 10px;
  width: 80vw;
`
const fadeIn = keyframes`
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
`

const Title = styled.h1`
  margin: auto;
  margin-bottom: 20px;
  color: #f6a623;
  font-size: 48px;
`

const Subtitle = styled.h2`
  margin: auto;
  margin-bottom: 25px;
  color: #90a4ae;
  font-size: 30px;
  font-weight: 300;
`

const AuthMessage = styled.h2.attrs({
  name: 'auth-message'
})`
  margin: auto;
  margin-bottom: 25px;
  color: #fe7867;
  font-size: 16px;
  animation: 0.3s ${fadeIn};
`

const BottomText = styled.div`
  font-size: 12px;
  color: #90a4ae;
  font-weight: 300;
  letter-spacing: 0.5px;
  margin-top: 16px;
`

const LoginButton = styled.button.attrs({
  type: 'submit'
})`
  width: 114px;
  height: 38px;
  padding: 5px;
  background-color: #cfd8dc;
  border-radius: 5px;
  color: white;
  font-size: 15px;
  font-family: Montserrat-Light;
  letter-spacing: 0.5px;
  border: none;
  margin-top: 15px;
  transition: opacity 1s;

  &:hover {
    opacity: 0.5;
  }

  *:active {
    border: 1px solid #cfd8dc;
    color: #cfd8dc;
    background-color: white;
  }
`

class LoginDetails extends React.Component {
  focusField = null

  static propTypes = {
    login: PropTypes.func.isRequired,
    push: PropTypes.func.isRequired,
    isAuthenticated: PropTypes.bool.isRequired,
    location: PropTypes.shape({
      state: PropTypes.shape({
        authMessage: PropTypes.string,
        from: PropTypes.shape({
          pathname: PropTypes.string
        })
      })
    }).isRequired
  }

  constructor(props) {
    super(props)
    this.state = {
      username: '',
      password: '',
      authMessage: ''
    }
  }

  componentWillMount() {}

  componentDidMount() {
    const { location } = this.props

    setTimeout(() => {
      if (this.focusField !== null) {
        this.focusField.focus()
      }
    }, 100)

    if (location && location.state && location.state.authMessage) {
      // Read the message from the previous route and display it (if applicable)
      const { authMessage } = location.state
      if (authMessage && authMessage !== '') this.setState({ authMessage })
    }
  }

  handleSubmit = event => {
    event.preventDefault()

    if (!this.state.username || !this.state.password) return

    const defaultRoute = '/app/dashboard'
    // If a referrer was specified, navigate to that route...
    const { location } = this.props
    let route =
      location.state && location.state.from && location.state.from.pathname
        ? location.state.from.pathname
        : defaultRoute

    // if (location.state.from.pathname)
    //   console.log('referrer found', location.state.from.pathname)

    this.props
      .login(this.state.username, this.state.password)
      .then(() => {
        this.props.push(route)
      })
      .catch(message => {
        this.setState({
          authMessage: message
        })
      })
  }

  handleUsernameChange = event => {
    this.setState({ username: event.target.value })
  }

  handlePasswordChange = event => {
    this.setState({ password: event.target.value })
  }

  render() {
    const { authMessage } = this.state
    return (
      <Container>
        <Title>Welcome to Zippy Operations</Title>
        <Subtitle>please log in to get started</Subtitle>
        <form onSubmit={this.handleSubmit}>
          {authMessage &&
            <AuthMessage>
              {authMessage}
            </AuthMessage>}
          <LoginInput
            name="username"
            value={this.state.username}
            placeholder="username"
            handleChange={this.handleUsernameChange}
            refFunc={field => (this.focusField = field)}
          />
          <LoginInput
            name="password"
            type="password"
            value={this.state.password}
            placeholder="password"
            handleChange={this.handlePasswordChange}
          />
          <LoginButton>Log In</LoginButton>
          <BottomText>
            Please log in with &quot;admin&quot; as username and password
          </BottomText>
        </form>
      </Container>
    )
  }
}

export default LoginDetails
