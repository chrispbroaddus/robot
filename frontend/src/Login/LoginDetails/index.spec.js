import React from 'react'
import { mount, shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import LoginDetails from './index'

describe('LoginDetails', () => {
  let props

  beforeEach(() => {
    props = {
      login: () => {},
      push: () => {},
      isAuthenticated: false,
      location: {
        state: {}
      }
    }
  })

  it('should render self and subcomponents', () => {
    const wrapper = shallow(<LoginDetails {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should set the authMessage state if a message was provided by a previous route', () => {
    const message = 'Test message'
    props.location = {
      state: {
        authMessage: message
      }
    }

    const wrapper = mount(<LoginDetails {...props} />)
    expect(wrapper.state('authMessage')).toBeTruthy()
  })

  it('should display the message from the previous route', () => {
    const message = 'Test message'
    props.location = {
      state: {
        authMessage: message
      }
    }

    const wrapper = mount(<LoginDetails {...props} />)
    const target = wrapper.find('h2[name="auth-message"]')
    expect(target.exists()).toBe(true)
    expect(target.text()).toEqual(message)
  })

  it('should display the login form', () => {
    const wrapper = mount(<LoginDetails {...props} />)
    const usernameField = wrapper.find('input[name="username"]')
    const passwordField = wrapper.find('input[name="password"]')
    const submit = wrapper.find('button[type="submit"]')
    const form = wrapper.find('form')
    expect(usernameField.exists()).toBe(true)
    expect(passwordField.exists()).toBe(true)
    expect(submit.exists()).toBe(true)
    expect(form.exists()).toBe(true)
  })

  it('should change the state when fields are changed', () => {
    const username = 'testUser'
    const password = 'myTestPassw0rd'

    const wrapper = mount(<LoginDetails {...props} />)
    const usernameField = wrapper.find('input[name="username"]')
    const passwordField = wrapper.find('input[name="password"]')
    usernameField.simulate('change', { target: { value: username } })
    passwordField.simulate('change', { target: { value: password } })

    expect(wrapper.state('username')).toEqual(username)
    expect(wrapper.state('password')).toEqual(password)
  })

  it('should call the login() prop when a username and password is provided', () => {
    props.login = jest.fn().mockReturnValue(
      new Promise(
        (resolve, reject) => {
          resolve()
        },
        (resolve, reject) => {
          reject()
        }
      )
    )
    props.push = jest.fn()
    const username = 'testUser'
    const password = 'myTestPassw0rd'

    const wrapper = mount(<LoginDetails {...props} />)
    const form = wrapper.find('form')
    wrapper.setState({
      username,
      password
    })
    form.simulate('submit')

    expect(props.login).toHaveBeenCalledWith(username, password)
  })

  it('should not call the login() prop when a username or password is missing', () => {
    props.login = jest.fn().mockReturnValue(
      new Promise(
        (resolve, reject) => {
          resolve()
        },
        (resolve, reject) => {
          reject()
        }
      )
    )
    props.push = jest.fn()
    const username = 'testUser'
    const password = 'myTestPassw0rd'

    const wrapper = mount(<LoginDetails {...props} />)
    const form = wrapper.find('form')

    // Password missing
    wrapper.setState({
      username,
      password: ''
    })
    form.simulate('submit')

    expect(props.login).not.toHaveBeenCalled()

    // Username missing
    wrapper.setState({
      username: '',
      password
    })
    form.simulate('submit')

    expect(props.login).not.toHaveBeenCalled()

    // Both missing
    wrapper.setState({
      username: '',
      password: ''
    })
    form.simulate('submit')

    expect(props.login).not.toHaveBeenCalled()
  })

  it('should not navigate when login fails', () => {
    const authMessage = 'failed login'
    props.push = jest.fn()
    props.login = jest.fn().mockReturnValue(
      new Promise((resolve, reject) => {
        expect(props.push).not.toHaveBeenCalled()
        reject(authMessage)
      })
    )

    const wrapper = mount(<LoginDetails {...props} />)
    const form = wrapper.find('form')
    wrapper.setState({
      username: 'testUser',
      password: 'myTestPassw0rd'
    })
    form.simulate('submit')

    expect.assertions(1)
  })

  it('should display the authMessage on fail', done => {
    const authMessage = 'failed login'
    props.push = jest.fn()
    props.login = jest.fn().mockReturnValue(
      new Promise((resolve, reject) => {
        expect(props.push).not.toHaveBeenCalled()
        reject(authMessage)
      })
    )

    const wrapper = mount(<LoginDetails {...props} />)
    const form = wrapper.find('form')
    wrapper.setState({
      username: 'testUser',
      password: 'myTestPassw0rd'
    })
    form.simulate('submit')

    expect.assertions(2)
    setTimeout(() => {
      expect(wrapper.state('authMessage')).toBe(authMessage)
      done()
    })
  })

  it('should navigate to the dashboard after a successful login', done => {
    props.login = jest.fn().mockReturnValue(
      new Promise((resolve, reject) => {
        resolve()
      })
    )
    props.push = jest.fn()

    const wrapper = mount(<LoginDetails {...props} />)
    const form = wrapper.find('form')
    wrapper.setState({
      username: 'testUser',
      password: 'myTestPassw0rd'
    })
    form.simulate('submit')

    expect.assertions(1)
    setTimeout(() => {
      expect(props.push).toHaveBeenCalledWith('/app/dashboard')
      done()
    })
  })

  it('should navigate to a custom route after a successful login', done => {
    const pathname = 'test/path/name'
    props.location = {
      state: {
        from: {
          pathname
        }
      }
    }
    props.login = jest.fn().mockReturnValue(
      new Promise((resolve, reject) => {
        resolve()
      })
    )
    props.push = jest.fn()

    const wrapper = mount(<LoginDetails {...props} />)
    const form = wrapper.find('form')
    wrapper.setState({
      username: 'testUser',
      password: 'myTestPassw0rd'
    })
    form.simulate('submit')

    expect.assertions(2)
    setTimeout(() => {
      expect(props.push).not.toHaveBeenCalledWith('/app/dashboard')
      expect(props.push).toHaveBeenCalledWith(pathname)
      done()
    })
  })
})
