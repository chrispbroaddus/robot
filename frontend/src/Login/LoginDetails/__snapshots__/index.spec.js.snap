// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginDetails should render self and subcomponents 1`] = `
<styled.div>
  <styled.h1>
    Welcome to Zippy Operations
  </styled.h1>
  <styled.h2>
    please log in to get started
  </styled.h2>
  <form
    onSubmit={[Function]}
  >
    <LoginInput
      handleChange={[Function]}
      name="username"
      placeholder="username"
      refFunc={[Function]}
      type="text"
      value=""
    />
    <LoginInput
      handleChange={[Function]}
      name="password"
      placeholder="password"
      type="password"
      value=""
    />
    <styled.button>
      Log In
    </styled.button>
    <styled.div>
      Please log in with "admin" as username and password
    </styled.div>
  </form>
</styled.div>
`;
